x-pack/filebeat/input/awss3/input.go

```diff
@@ -62,13 +94,55 @@ func (in *s3Input) Test(ctx v2.TestContext) error {
 }
 
 func (in *s3Input) Run(ctx v2.Context, pipeline beat.Pipeline) error {
-	collector, err := in.createCollector(ctx, pipeline)
+	var err error
+	var collector s3Collector
+
+	persistentStore, err := in.store.Access()
+	if err != nil {
+		return fmt.Errorf("Can not access persistent store: %w", err)
+	}
+
+	states := NewStates()
+	err = states.readStatesFrom(persistentStore)
+	if err != nil {
+		return fmt.Errorf("Can not start persistent store: %w", err)
+	}
+
+	metricRegistry := monitoring.GetNamespace("dataset").GetRegistry()
+	inputMetrics := newInputMetrics(metricRegistry, ctx.ID)
+
+	if in.config.QueueURL != "" {
+		collector, err = in.createSQSCollector(ctx, pipeline, inputMetrics)
+		if err != nil {
+			return fmt.Errorf("cannot create SQS collector: %w", err)
+		}
+	}
+
+	if in.config.S3Bucket != "" {
+		collector, err = in.createS3BucketCollector(ctx, pipeline, inputMetrics, states, persistentStore)
+		if err != nil {
+			return fmt.Errorf("cannot create S3 bucket collector: %w", err)
+		}
+	}
+
+	err = in.grp.Go(func(canceler unison.Canceler) error {
+		interval := in.store.CleanupInterval()
+		if interval <= 0 {
+			interval = 5 * time.Minute
+		}
+
+		interval = 1 * time.Minute
+		cleanStore(canceler, ctx.Logger, persistentStore, states, interval)
```

Comment:
I've been thinking about this cleanup logic, can we really separate bucket listing from cleanup? If we use `lastModified` dates for cleanup, somehow we need to guarantee that cleanup happens **immediately after** a full listing, otherwise we are at risk of skipping not-yet-listed objects that have a `lastModified` time older than what we saw so far.

Reply:
I see the problem, but I think then that to solve this we should bind cleanup not to bucket listing but to the mark-as-stored step.
The problem I see is that mark-as-stored is done at object level, and not for the whole listing: and still we will end up in the same situation where we skip not-yet-listed objects that have a `lastModified` older than what we marked in the mark-as-stored step.

We are trying to infer a state (or the outcome of a missing state: skip the object) from a state that's missing in store.

Another option is to keep states by listing operation, and remove the whole batch at once when all of them are processed/marked-as-stored. In this case we have a precise batch we know that either we have a state for we can compare to, or we miss and we should not skip. What do you think?

Reply:
> Another option is to keep states by listing operation, and remove the whole batch at once when all of them are processed/marked-as-stored. In this case we have a precise batch we know that either we have a state for we can compare to, or we miss and we should not skip. What do you think?

This sounds reasonable to me, a few concerns:
- A later list operation should also be able to invalidate older states (right?)
- Would memory usage be an issue? I think it wouldn't but may be missing something

Reply:
>A later list operation should also be able to invalidate older states (right?)

I was thinking about keeping "buckets" of list operations, once all the state in a bucket are marked as stored the bucket will be purged. When we check if an object can be skipped we check in every buckets of the list operations.
We could reduce to O(n) (n = number of list operations) with the right algorithm

> Would memory usage be an issue? I think it wouldn't but may be missing something

Memory wise we will increase the requirements by a factor of `s3_bucket_poll_interval/cleanup_interval`. Since the default values are respectively 120s and 300s we will something between 2x and 3x

Reply:
> I was thinking about keeping "buckets" of list operations, once all the state in a bucket are marked as stored the bucket will be purged. When we check if an object can be skipped we check in every buckets of the list operations.

Can 2 listings happen at the same time? They are run in loop, keeping the state for all of them presents some issues, for instance, interrupted listings (for instance if FB dies). I'm wondering if it would be just fine to care about the last successful listing. Always doing a cleanup after a it, removing previous states.